R/percent.R

```diff
@@ -0,0 +1,194 @@
+}
+signif_half_up <- function(x, digits = 6){
+  if (is.null(digits) || (length(digits) == 1 && digits == Inf)){
+    return(x)
+  }
+  round_half_up(x, digits - ceiling(log10(abs(x))))
+}
+
+#' @export
+as.character.percent <- function(x, digits = 2, ...){
+  if (length(x) == 0){
+    character()
+  } else {
+    paste0(unclass(round(x, digits) * 100), "%")
+  }
+}
+
+#' @export
+format.percent <- function(x, symbol = "%", trim = TRUE,
+                           digits = 2,
+                           ...){
+  if (length(x) == 0){
+    out <- character()
+  } else {
+    out <- paste0(format(unclass(round(x, digits) * 100), trim = trim, digits = NULL, ...),
+                  symbol)
+  }
+  names(out) <- names(x)
+  out
+}
+
+#' @export
+print.percent <- function(x, max = NULL, trim = TRUE,
+                          digits = 2,
+                          ...){
+  out <- x
+  N <- length(out)
+  if (N == 0){
+    print("percent(numeric())")
+    return(invisible(x))
+  }
+  if (is.null(max)) {
+    max <- getOption("max.print", 9999L)
+  }
+  suffix <- character()
+  max <- min(max, N)
+  if (max < N) {
+    out <- out[seq_len(max)]
+    suffix <- paste(" [ reached 'max' / getOption(\"max.print\") -- omitted",
+                    N - max, "entries ]\n")
+  }
+  print(format(out, trim = trim, digits = digits), ...)
+  cat(suffix)
+  invisible(x)
+}
+
+#' @export
+`[.percent` <- function(x, ..., drop = TRUE){
+  cl <- oldClass(x)
+  class(x) <- NULL
+  out <- NextMethod("[")
+  class(out) <- cl
+  out
+}
+
+#' @export
+unique.percent <- function(x, incomparables = FALSE,
+                           fromLast = FALSE, nmax = NA, ...){
+  cl <- oldClass(x)
+  class(x) <- NULL
+  out <- NextMethod("unique")
+  class(out) <- cl
+  out
+}
+
+#' @export
+rep.percent <- function(x, ...){
+  cl <- oldClass(x)
+  class(x) <- NULL
+  out <- NextMethod("rep")
+  class(out) <- cl
+  out
+}
+
+#' @export
+Ops.percent <- function(e1, e2){
+  math <- switch(.Generic,
+                 `+` =,
+                 `-` =,
+                 `*` =,
+                 `/` =,
+                 `^` =,
+                 `%%` =,
+                 `%/%` = TRUE, FALSE)
+  if (inherits(e2, "percent") && !inherits(e1, "percent")){
+    e1 <- unclass(e1)
+    e2 <- unclass(e2)
+  }
+  NextMethod(.Generic)
+}
+#' @export
+Math.percent <- function(x, ...){
+  rounding_math <- switch(.Generic,
+                          `floor` =,
+                          `ceiling` =,
+                          `trunc` =,
+                          `round` =,
+                          `signif` = TRUE, FALSE)
+  x <- unclass(x)
+  if (rounding_math){
+    x <- x * 100
+    if (.Generic == "round"){
+      out <- do.call(round_half_up, list(x, ...))
+    } else if (.Generic == "signif"){
+      out <- do.call(signif_half_up, list(x, ...))
+    } else {
+      out <- NextMethod(.Generic)
+    }
+    new_percent(out / 100)
+  } else {
+    out <- NextMethod(.Generic)
+    new_percent(out)
+  }
+}
```

Comment:
Do we need additional methods for `mean`, `median`, `sd`, `quantile`..?

Reply:
Yes I can add methods for the functions defined in the `Summary` group generic.
I'll also add `mean.percent()`.
These should handle everything except `sd()` and `var()` which I don't think you can write methods for though I can live without this for now.

Reply:
`abs()` works, basically all the Math, Summary and Ops group generics should work as expected, see ?Math for more details..

Reply:
`var`, `sd` and `mean` aren't included in those generics which makes sense as they're part of the stats package.

Reply:
But at least mean is an S3 generic so we can easily write a method for it, which I have done in the latest commits.